### ISSUE TITLE
[React][Dialog] Controlled Dialog is not fully controlled

### ISSUE THREAD

**Issue by fseglard** [chakra-ui/ark#1718]
### Description

Hi!

When using the `Dialog` component in controlled mode, there is an issue with the close behavior. When in controlled mode, I expect to fully control "how" the dialog will be closed but that's not the case currently as the dialog will always be closed internally when clicking on the `CloseTrigger` component, using the Esc key, or clicking on the backdrop.

As a workaround, I could listen for the `onOpenChange` property and update the parent state, but it does not make any sense as  the open state of the dialog should be "controlled" by the parent directly. Also, it causes an extra render because we're trying to sync both the state of the parent and the internal dialog state.

When looking at other component libraries (MUI, HeadlessUI, etc.), a dialog in controlled mode won't be closed unless the parent sets the proper `open` value.

### Link to Reproduction (or Detailed Explanation)

https://codesandbox.io/s/ark-ui-dialog-controlled-2l28fg?file=/src/Dialog.tsx

### Steps to Reproduce

1. Go to the reproduction link
2. Open the dialog
3. Close the dialog by clicking on the close button
4. Check the console logs, you should see a message indicating that the `onOpenChange` hander has been called with the `false` value meaning the internal dialog state has changed
5. Try to open the dialog again
6. The dialog won't open because the parent state is not in sync.

**Expected behavior**
The dialog should not close in this case when I click on the close button because I don't set the `open` props to `false`.

### Ark UI Version

1.0.1

### Framework

- [X] React
- [ ] Solid
- [ ] Vue

### Browser

Google Chrome 119

### Additional Information

It's possible that we have the same issue with other components that have a controlled mode and overlay.

**Comment by cschroeter** [comment]
@fseglard 

Mhh I'm a bit indifferent on this one. If you want to control a component you should also add the `onOpenChange` handler to your component, similiar when you would use an `<input value="foo" />` and get this warning: `Warning: You provided a `value` prop to a form field without an `onChange` handler.`

Also would learn the motivation for this behaviour. What do you like to achieve with this?

I will sync with @segunadebayo on this one.

**Comment by fseglard** [comment]
@cschroeter 

I totally agree that you should have the `onChange` handler (or the `onOpenChange` handler in the case of the dialog) and this is currently the case if you look at the codesandbox example. But that's not the point of the issue.

The example you gave with the input is an excellent one. When you "control" an input, the value of the input is controlled by the parent component using the `onChange` handler. So the value of the input will never change unless the parent set a proper value.

In the case of the dialog, I expect the same behavior. The dialog should not be closed unless the parent specifically pass the `open` props to `false`.

My personal use case is that I need to trigger a confirmation message for the user when he wants to close the dialog. If the user confirm that he wants to close, then the dialog should close, but if he doesn't, the modal should stay open.

Hope this helps!

**Comment by cschroeter** [comment]
@fseglard 

Understood your usecase, but I don't understand why you need to control the dialog. See example: https://codesandbox.io/s/ark-ui-dialog-controlled-forked-jdrvyq?file=/src/Dialog.tsx

**Comment by fseglard** [comment]
Yes, I could use some UI/UX alternatives to do that, but I still think that this is an issue in Ark-UI as the controlled mode doesn't work exactly how it should be.

Also in your example, you disable the `closeOnEscapeKeyDown` and `closeOnInteractOutside` but I don't want to disable them! The `onOpenChange` handler should be called when the user click on the backdrop, and let the parent handle the logic. Otherwise, I don't see the point to have a controlled mode.

**Comment by cschroeter** [comment]
> Otherwise, I don't see the point to have a controlled mode.

The reason this make sense is when you want to control the `Dialog` from outside and don't want to rely on the `DialogTrigger` within the `Dialog`.

Anyhow this is not supported at the moment in Zag.js so let me transfer this issue over there.
